lib/Service/ReverseGeoCoderService.php

```diff
@@ -0,0 +1,167 @@
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU Affero General Public License as
+ * published by the Free Software Foundation, either version 3 of the
+ * License, or (at your option) any later version.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU Affero General Public License for more details.
+ *
+ * You should have received a copy of the GNU Affero General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+namespace OCA\Photos\Service;
+
+use OCP\Files\IAppData;
+use OCP\Files\SimpleFS\ISimpleFolder;
+use OCP\Files\NotFoundException;
+use OCP\Http\Client\IClientService;
+use Hexogen\KDTree\FSTreePersister;
+use Hexogen\KDTree\FSKDTree;
+use Hexogen\KDTree\KDTree;
+use Hexogen\KDTree\Item;
+use Hexogen\KDTree\ItemList;
+use Hexogen\KDTree\ItemFactory;
+use Hexogen\KDTree\NearestSearch;
+use Hexogen\KDTree\Point;
+
+class ReverseGeoCoderService {
+	private IClientService $clientService;
+	private ISimpleFolder $geoNameFolder;
+	private ?NearestSearch $fsSearcher = null;
+	/** @var array<int, string> */
+	private ?array $citiesMapping = null;
+
+	public function __construct(
+		IAppData $appData,
+		IClientService $clientService
+	) {
+		$this->clientService = $clientService;
+
+		try {
+			$this->geoNameFolder = $appData->getFolder("geonames");
+		} catch (\Exception $ex) {
+			if ($ex instanceof NotFoundException) {
+				$this->geoNameFolder = $appData->newFolder("geonames");
+			}
+
+			throw $ex;
+		}
+	}
+
+	public function getLocationIdForCoordinates(float $latitude, float $longitude): int {
+		if ($this->fsSearcher === null) {
+			$this->buildKDTree();
+			$kdTreeFileContent = $this->geoNameFolder->getFile("cities1000.bin")->getContent();
+			$kdTreeTmpFileName = tempnam(sys_get_temp_dir(), "nextcloud_photos_");
+			file_put_contents($kdTreeTmpFileName, $kdTreeFileContent);
+			$fsTree = new FSKDTree($kdTreeTmpFileName, new ItemFactory());
+			$this->fsSearcher = new NearestSearch($fsTree);
+		}
+
+		$result = $this->fsSearcher->search(new Point([$latitude, $longitude]), 1);
+		return $result[0]->getId();
+	}
+
+	public function getLocationNameForLocationId(int $locationId): string {
+		if ($this->citiesMapping === null) {
+			$this->downloadCities1000();
+			$cities1000 = $this->loadCities1000();
+			foreach ($cities1000 as $city) {
+				$this->citiesMapping[$city['id']] = $city['name'];
+			}
+		}
+
+		return $this->citiesMapping[$locationId] ?? '';
+	}
+
+	private function downloadCities1000(bool $force = false) {
+		if ($this->geoNameFolder->fileExists('cities1000.csv') && !$force) {
+			return;
+		}
+
+		// Download zip file to a tmp file.
+		$response = $this->clientService->newClient()->get("http://download.geonames.org/export/dump/cities1000.zip");
```

Comment:
Just wanted to drop a note FYI: this database doesn't work very well if you query the closest data point, at least in urban US. This is mostly due to the large number of cities in metropolitan regions (I don't know what the case is in Europe / other countries). 

For instance, querying the closest location to me in the DB would either be Santa Monica or Westwood but I'm actually in LA. LA has 88 cities (and more unincorporated ones / villages in the DB), so photos from around the city would likely create a mess.

@artonge

Reply:
Ah, I thought I answered you already.

What would you recommend ?

Reply:
Hmm GitHub doesn't seem to notify correctly on review comments 🤔 

I've been trying to find a good solution to this for a long time now, but there's no easy one. There's no freely available dataset with cities and their boundaries that can be used directly. One idea is to use Overpass to dump OSM data into such a dataset (could use Geonames to get the list of cities to dump). Querying this efficiently might also be a bit challenging. I don't expect the dataset to be very large though; maybe a few gigabytes at most.

For now, I do think using Geonames is a good start, so I'll recommend going ahead with this 👍🏻 
Just wanted to drop this caution here in case someone can come up with a better idea.

Reply:
:+1: thanks for sharing. Indeed, I don't see any other solution.

Reply:
@artonge so I did finally get around to implementing this. The solution I settled on was to extract and simplify boundary data from the OSM planet database, then do an indexed db query on it. The results are _very_ accurate and the total data download size is at a reasonable ~80MB. Queries on this are surprisingly fast; on my slow-ish server it takes just ~1ms to query the entire planet. The downside is SQLite doesn't support geometric functions so it won't work there (but then who runs SQLite in production anyway).

I wrote a post about it if you're curious: https://varunpp2.medium.com/nextcloud-memories-putting-thousands-of-photos-on-the-world-map-1376cf4105a0